specs/bellatrix/fork-choice.md

```diff
@@ -78,6 +79,100 @@ As per EIP-3675, before a post-transition block is finalized, `notify_forkchoice
+`validator_index` is connected to the node (e.g. has sent an unexpired proposer preparation
+message).
+
+```python
+def validator_is_connected(validator_index: ValidatorIndex) -> boolean:
+    ...
+```
+
+```python
+def should_override_forkchoice_update(
+    store: Store,
+    head_root: Root,
+) -> boolean:
+    justified_state = store.checkpoint_states[store.justified_checkpoint]
+    head_block = store.blocks[head_root]
+    parent_root = head_block.parent_root
+    parent_block = store.blocks[parent_root]
+    current_slot = get_current_slot(store)
+
+    # Single slot re-org.
+    parent_slot_ok = parent_block.slot + 1 == head_block.slot
+    time_into_slot = (store.time - store.genesis_time) % SECONDS_PER_SLOT
+    current_time_ok = (head_block.slot == current_slot or
+                       (head_block.slot + 1 == current_slot and
+                        time_into_slot < SECONDS_PER_SLOT // INTERVALS_PER_SLOT))
+    single_slot_reorg = parent_slot_ok and current_time_ok
+
+    # Shuffling stable.
+    shuffling_stable = (head_block.slot + 1) % SLOTS_PER_EPOCH != 0
+
+    # FFG information of the new head_block will be competitive with the current head.
+    # FIXME: needs more justification/finalization information in the store
+    # ffg_competitive =
+    #    store.realized_justifications[parent_root] ==
+    #        store.realized_justifications[head_root] and
+    #    store.realized_finalizations[parent_root] ==
+    #        store.realized_finalization[head_root]
+    ffg_competitive = True
+
+    # Ensure that total participation meets a minimum threshold.
+    # The participation multiplier is set to the number of slots for which the parent_block block has
+    # been known.
+    participation_multiplier = current_slot - parent_block.slot
+    parent_weight = get_latest_attesting_balance(store, parent_root)
+    participation_threshold = calculate_committee_fraction(justified_state, PARTICIPATION_THRESHOLD)
+    participation_ok = parent_weight >= participation_multiplier * participation_threshold
+
+    # Check the head weight only if the attestations from the head slot have already been applied.
+    # Implementations may want to do this in different ways, e.g. by advancing
+    # `store.time` early, or by counting queued attestations during the head block's slot.
+    if current_slot > head_block.slot:
+        assert store.proposer_boost_root == Root()
+        head_weight = get_latest_attesting_balance(store, head_root)
+        reorg_threshold = calculate_committee_fraction(justified_state, REORG_WEIGHT_THRESHOLD)
+        head_weak = head_weight < reorg_threshold
+    else:
+        head_weak = True
+
+    # Only re-org the head_block block if it arrived later than the attestation deadline.
+    head_late = store.block_timeliness.get(head_root) is False
+
+    # Only suppress the fork choice update if we are confident that we will propose the next block.
```

Comment:
My naive intuition is this should be the very first check and just exit early if no one is proposing the next block : )

Reply:
I think the optimal implementation returns early as soon as any condition is false, and does the checks in order from cheapest to most expensive (which might depend on the client architecture + cache structure). For the spec I don't mind trying to roughly match it to the clients, but I think there'll necessarily be a degree of divergence, particularly if we leave out the complexity of the early returns (which I think we should).

Maybe an intuitive order is:

1. `block_late`
2. `proposing_next_slot`
3. `single_slot_reorg`, `shuffling_stable`, `ffg_competitive`, `participation_ok`, `head_weak`

And I can also add a note suggesting that clients implement the checks in an order that makes sense for them, with appropriate early returns?

Reply:
Updated

Reply:
Verygood👌🌹🙏